tests/unit/s2n_self_talk_ktls_test.c

```diff
@@ -321,19 +321,138 @@ int main(int argc, char **argv)
+        {
+            const uint8_t test_record_type = TLS_CHANGE_CIPHER_SPEC;
+            uint8_t control_record_data[] = "control record data";
+            struct s2n_blob control_record = { 0 };
+            EXPECT_SUCCESS(s2n_blob_init(&control_record, control_record_data,
+                    sizeof(control_record_data)));
+
+            for (size_t i = 0; i < 5; i++) {
+                EXPECT_OK(s2n_record_write(writer, test_record_type, &control_record));
+                EXPECT_SUCCESS(s2n_flush(writer, &blocked));
+
+                int written = s2n_send(writer, test_data, sizeof(test_data), &blocked);
+                EXPECT_EQUAL(written, sizeof(test_data));
+
+                uint8_t buffer[sizeof(test_data)] = { 0 };
+                int read = s2n_recv(reader, buffer, sizeof(buffer), &blocked);
+                EXPECT_EQUAL(read, sizeof(test_data));
+                EXPECT_EQUAL(blocked, S2N_NOT_BLOCKED);
+
+                EXPECT_BYTEARRAY_EQUAL(test_data, buffer, read);
+            }
+        };
 
-        /* Test: s2n_recv not implemented yet */
+        /* Test: s2n_recv with incorrectly encrypted application data
+         *
+         * This test closes the connection so should be the last test to use
+         * these connections.
+         */
         {
-            uint8_t buffer[10] = { 0 };
-            int received = s2n_recv(reader, buffer, sizeof(buffer), &blocked);
-            EXPECT_FAILURE_WITH_ERRNO(received, S2N_ERR_UNIMPLEMENTED);
+            /* Write a valid record of application data */
+            EXPECT_OK(s2n_record_write(writer, TLS_APPLICATION_DATA, &test_data_blob));
+            /* Wipe part of the encrypted record so that it is no longer valid */
+            EXPECT_SUCCESS(s2n_stuffer_wipe_n(&writer->out, 10));
+            EXPECT_SUCCESS(s2n_stuffer_skip_write(&writer->out, 10));
+            /* Actually send the modified record */
+            EXPECT_SUCCESS(s2n_flush(writer, &blocked));
+
+            uint8_t buffer[sizeof(test_data)] = { 0 };
+            int read = s2n_recv(reader, buffer, sizeof(buffer), &blocked);
+            EXPECT_FAILURE_WITH_ERRNO(read, S2N_ERR_IO);
+            EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_READ);
+
+            /* This error is fatal and blinded */
+            EXPECT_TRUE(s2n_connection_check_io_status(reader, S2N_IO_CLOSED));
+            EXPECT_TRUE(s2n_connection_get_delay(reader) > 0);
+            EXPECT_TRUE(s2n_connection_get_delay(reader) < UINT64_MAX);
+        };
+    };
+
+    /* Test: s2n_shutdown
+     *
+     * There are three ways to trigger the read side of a TLS connection to close:
+     * 1. Receive an alert while calling s2n_recv
+     * 2. Receive an alert while calling s2n_shutdown
+     * 3. Receive "end of data" while calling s2n_recv (but this is an error)
+     *
+     * We need a fresh connection to test each scenario. Wiping a connection isn't
+     * sufficient because we can't disable / reset ktls.
```

Comment:
I might be misunderstanding - isn't this test wiping the connection for each scenario?

Reply:
You're not misunderstanding, that's just a horribly written comment. Fixed.